instrumentation/github.com/aws/aws-lambda-go/otellambda/lambda.go

```diff
@@ -0,0 +1,351 @@
+// Get returns the value associated with the passed key.
+func (mc mapCarrier) Get(key string) string {
+	return mc[key]
+}
+
+// Set stores the key-value pair.
+func (mc mapCarrier) Set(key string, value string) {
+	mc[key] = value
+}
+
+// Keys lists the keys stored in this carrier.
+func (mc mapCarrier) Keys() []string {
+	keys := make([]string, len(mc))
+	i := 0
+	for k := range mc {
+		keys[i] = k
+		i++
+	}
+	return keys
+}
+
+// Adds OTel span surrounding customer handler call
+func wrapper(handlerFunc interface{}) func(ctx context.Context, event interface{}, takesContext bool) []reflect.Value {
+	return func(ctx context.Context, event interface{}, takesContext bool) []reflect.Value {
+
+		ctx, span := tracingBegin(ctx)
+		defer tracingEnd(ctx, span)
+
+		handler := reflect.ValueOf(handlerFunc)
+		var args []reflect.Value
+		if takesContext {
+			args = append(args, reflect.ValueOf(ctx))
+		}
+		if eventExists(event) {
+			args = append(args, reflect.ValueOf(event))
+		}
+
+		response := handler.Call(args)
+
+		return response
+	}
+}
+
+// Determine if an interface{} is nil or the
+// if the reflect.Value of the event is nil
+func eventExists(event interface{}) bool {
+	if event == nil {
+		return false
+	}
+
+	// reflect.Value.isNil() can only be called on
+	// Values of certain Kinds. Unsupported Kinds
+	// will panic rather than return false
+	switch reflect.TypeOf(event).Kind() {
+	case reflect.Chan, reflect.Func, reflect.Map, reflect.Ptr, reflect.UnsafePointer, reflect.Interface, reflect.Slice:
+		return !reflect.ValueOf(event).IsNil()
+	}
+	return true
+}
+
+type wrappedHandler struct {
+	handler lambda.Handler
+}
+
+// Compile time check our Handler implements lambda.Handler
+var _ lambda.Handler = wrappedHandler{}
+
+// Invoke adds OTel span surrounding customer Handler invocation
+func (h wrappedHandler) Invoke(ctx context.Context, payload []byte) ([]byte, error) {
+
+	ctx, span := tracingBegin(ctx)
+	defer tracingEnd(ctx, span)
+
+	response, err := h.handler.Invoke(ctx, payload)
+	if err != nil {
+		return nil, err
+	}
+
+	return response, nil
+}
+
+// HandlerWrapper Provides a Handler which wraps customer Handler with OTel Tracing
+func HandlerWrapper(handler lambda.Handler) lambda.Handler {
+	return wrappedHandler{handler: handler}
+}
+
+// Logic to start OTel Tracing
+func tracingBegin(ctx context.Context) (context.Context, trace.Span) {
+	// Add trace id to context
+	xrayTraceId := os.Getenv("_X_AMZN_TRACE_ID")
+	mc := mapCarrier{}
+	mc.Set("X-Amzn-Trace-Id", xrayTraceId)
+	propagator := xray.Propagator{}
+	ctx = propagator.Extract(ctx, mc)
+
+	// If tracer provider initialization failed we
+	// will attempt to initialize once per invocation
+	if tp == nil {
+		initTracerProvider()
+	}
+
+	// if tracer provider successfully initializes then
+	// we add tracing, otherwise do customer business
+	// logic with no tracing
+	if tp != nil {
+		// Get a named tracer with package path as its name.
+		tracer := tp.Tracer(tracerName)
```

Comment:
It would be preferable to have this function be a method hanging off a type like `wrappedHandler` that could retain a reference to the `Tracer`.  The value of `tracerName` doesn't change and there is no need to obtain a new `Tracer` with every invocation.

Reply:
Now stores a Tracer to use each invocation